src/yellowpaper/chapters/03-implementation/data-disclosure.tex

```diff
@@ -1,64 +1,83 @@
+\item \tParty{E} transfers the encrypted data chunks $\Lock{D}$ directly to the buyer. \tParty{B} computes the Merkle root $R'$ and the size $s'$ of the received data $\Lock{D}'$:
 \begin{equation}
-R' = \Root(\MerkleTree(\Lock{E}'))
+R' = \Root(\MerkleTree(\Lock{D}'))
 \end{equation}
 \begin{equation}
-s' = \SizeOf(\Lock{E}')
+s' = \SizeOf(\Lock{D}')
 \end{equation}
-\item \tParty{B} makes a transaction with a receipt $\Sign_\Party{B}(\{R',\ s'\})$ to the contract address. The parties can proceed if and only if the following is true:
+\item \label{step:buyers-receipt} \tParty{B} makes a transaction with a receipt $\Sign_\Party{B}(\{R',\ s'\})$ to the contract address. The parties can proceed if and only if the following is true:
 \begin{equation}
 (R' = R)\ \land\ (s' = s)
 \end{equation}
 Otherwise, the protocol halts.
-\item \tParty{E} sends $\Sign_\Party{E}(\Encrypt_\Party{B}(k))$ to the contract.
-\item \tParty{B} decyphers and checks the received data. In case some data chunk $e_i \in \Lock{E}$ is invalid, \tParty{B}~sends a transaction with~$\{\SecretKey{B},\ e_i,\ \Path(e_i,\ \MerkleTree(\Lock{E})),\ \Path(d_i,\ \MerkleTree(\Lock{D}))\}$ to the contract. By doing so, \tParty{B}~reveals the data chunk~$d_i$ corresponding to the encrypted chunk~$e_i$.  She also shares proof that~$e_i$ was indeed part of a Merkle tree with root~$R$ and the~$d_i$ was part of a Merkle tree~$R_0$. The contract checks the validity of $d_i$ and decides whether \tParty{B}~has rightfully accused~\tParty{E} of cheating.
+\item \label{step:secret-sharing} \tParty{E} sends $\Sign_\Party{E}(\Encrypt_\Party{B}(k))$ to the contract.
+\item \label{step:arbitrage} \tParty{B} decyphers and checks the received data.
+
+  \begin{itemize}
+    \item In case some data chunk $e_i \in \Lock{D}$ is invalid, \tParty{B}~sends
+    \begin{equation*}
+      \Sign_\Party{B}(\{\ \SecretKey{B},\ e_i,\ \Path(e_i,\ \MerkleTree(\Lock{D})),\ \Path(d_i,\ \MerkleTree(D))\ \})
+    \end{equation*}
+    to the contract. By doing so, \tParty{B}~reveals the data chunk~$d_i$ corresponding to the encrypted chunk~$e_i$. She also shares proof that~$e_i$ was indeed part of a  Mekle tree with root~$R$. The contract checks the validity of $d_i$ and decides whether \tParty{B}~has rightfully accused~\tParty{E} of cheating.
+
+    \item In case chunks $d_i$ and $d_j$ have duplicate entries, \tParty{B} sends
+    \begin{equation*}
+      \begin{split}
+        \Sign_\Party{B}(\{\ \SecretKey{B}, & \ e_i,\ \Path(e_i,\ \MerkleTree(\Lock{D})),\ \Path(d_i,\ \MerkleTree(D)),\\
+         & \ e_j,\ \Path(e_j,\ \MerkleTree(\Lock{D})),\ \Path(d_j,\ \MerkleTree(D))\ \})
+      \end{split}
+    \end{equation*}
+    to the contract. The contract checks whether $d_i$ and $d_j$ do indeed have duplicate entries and blames~\tParty{E} for cheating if it is true.
+  \end{itemize}
```

Comment:
What if the data was correct? Shouldn't we mention that case and describe the message send to the contract, so it can send money to `E` immidiately?